Companion-API oauth register/update user

blancinot opened the pull request.

Commit:
feat(oauth): companion-api can register new user w/ oauth

Commit:
chore(): merge

Commit:
fix(ws): same root path

Commit:
feat(oauth): register oauth acess token too

Commit:
fix(oauth): better w/ code 200 & auth header

mcluseau commented on pkg/companion-api/api/oauth.go:
```diff
@@ -0,0 +1,155 @@
+func setup() {
+	setupOauthOnce.Do(func() {
+		oauthConfig = &oauth2.Config{
+			RedirectURL:  requireEnv("OAUTH_MYAPP_ENDPOINT", "This app endpoint") + "/oauth/callback",
```
`OAUTH_APP_ENDPOINT` sounds better to me

mcluseau commented on pkg/companion-api/api/oauth.go:
```diff
@@ -0,0 +1,155 @@
+	restful "github.com/emicklei/go-restful"
+	"github.com/mcluseau/autentigo/auth"
+	"github.com/mcluseau/autentigo/pkg/companion-api/backend"
+	"golang.org/x/oauth2"
```
care imports ordering

mcluseau commented on pkg/companion-api/api/oauth.go:
```diff
@@ -0,0 +1,155 @@
+		return
+	}
+
+	r, errr := http.Get(oauthUserInfosURL + "?fields=name,email&access_token=" + token.AccessToken)
```
to many `r`s

blancinot replied:
rrrrrright :D

mcluseau commented on pkg/companion-api/backend/client.go:
```diff
@@ -7,6 +7,7 @@ import (
 // UserData is a simple user struct with paswordhash and claims
 type UserData struct {
 	PasswordHash string           `json:"password"`
+	OauthToken   string           `json:"oauth_token"`
```
I'd expected multiple oauth identities bound to a single account here (ie: 1 google, 1 github, etc)

blancinot replied:
ok. It means that for each client account (ie: 1 google, 1 github, etc.), 1 companion-api will have to be deployed. Or it is also necessary that the companion-api can manage the "n-account" with several CLIENTID, ... etc?

mcluseau replied:
I'd say we want something like `<prefix>/oauth/<provider>/<client-id>` pointing to the associated user account.

mcluseau replied:
what I mainly meant was to use this path as an *etcd* storage key, ie associating user `foo` with clientid `bar` in provider `baz` is `kv.Put("oauth/baz/bar", "foo")` (and the symmetric `Get` to check a clientid knowning the source provider to get the user id to query).

mcluseau replied:
but the multiprovider support is also required and LGTM ;)

Commit:
fix(oauth): something like <prefix>/oauth/<provider> pointing to the associated user account.

blancinot commented:
second (better) try :smile:

Commit:
fix(oauth): care imports ordering

Commit:
feat(oauth): association userID with clientID

feat(oauth): association userID with clientID (clean)

feat(oauth): association userID with clientID (clean)
